[amp-story] Prerender API

Enriqe opened the pull request:
Closes #31900
Closes #30513

Exposes a method which can be used to prerender a story. When desired, `play()` can be called to start playing the story.


It also adds stories to the player if they are missing from when calling `show()`
<!--
# Instructions:

- Pick a meaningful title for your pull request. (Use sentence case.)
  - Prefix the title with an emoji to identify what is being done. (Copy-paste the emoji from the list below.)
  - Do not overuse punctuation in the title (like `(chore):`).
  - If it is helpful, use a simple prefix (like `ProjectX: Implement some feature`).
- Enter a succinct description that says why the PR is necessary, and what it does.
  - Mention the GitHub issue that is being addressed by the pull request.
  - The keywords `Fixes`, `Closes`, or `Resolves` followed the issue number will automatically close the issue.

> NOTE: All non-trivial changes (like introducing new features or components) should have an associated issue or reference an I2I (intent-to-implement: go.amp.dev/i2i). Please read through the contribution process (go.amp.dev/contributing/code) for more information.

# Example of a good description:

- Implement aspect X
- Leave out feature Y because of A
- Improve performance by B
- Improve accessibility by C

# Emojis for categorizing pull requests (copy-paste emoji into description):

✨ New feature
🐛 Bug fix
🔥 P0 fix
✅ Tests
❄️ Flaky tests
🚀 Performance improvements
🖍 CSS / Styling
♿ Accessibility
🌐 Internationalization
📖 Documentation
🏗 Infrastructure / Tooling / Builds / CI
⏪ Reverting a previous change
♻️ Refactoring
🚮 Deleting code
🧪 Experimental code
-->

mszylkowski commented on src/amp-story-player/amp-story-player-impl.js:
```diff
@@ -302,6 +303,16 @@ export class AmpStoryPlayer {
+   */
+  prerender(storyUrl) {
+    this.autoplay_ = false;
+    this.show(storyUrl).then(() => (this.autoplay_ = true));
```
If someone calls `prerender` but have the config `autoplay: false`, shouldn't it keep that behavior?

Enriqe replied:
Yep good point, we should probably save the autoplay state to return it to.

> Also, what's the difference between this and just calling show?

It loads a story on the player on prerender mode without playing it. It's basically syntactic sugar for opting out of autoplay and calling show().

An example would be a carousel/grid with multiple stories. The publisher would like to prerender certain story since they know it's the most likely to be clicked. They can just call `player.prerender(storyUrl)` which will start loading the story without playing it, and then when the user actually clicks on something, they'll call `player.play()` to start playing the story.

mszylkowski replied:
Ok, then I think what you have now is good, since it would start autoplaying when you scroll to it right?

mszylkowski reviewed (approved):
LGTM

gmajoulet reviewed (commented):
I don't think re-using `show()` will work as it does more than prerendering. The fact that we have to hack around the `autoplay` is a big sign that it doesn't work and will have side effects.
The current implementation would break many scenarios developers would expect to work, for example:

- Prerendering a story while the Player is visible (as it navigates to the story when prerendering)
- Prerendering multiple stories will change the `autoplay` setting

Enriqe commented:
Yeah, i was having trouble deciding what exactly the `prerender` API should do and assumed it meant preloading the story in the displayed position of the player but not play it.

I guess your suggestion of prerendering stories in the background makes more sense :p. Then publishers would just need to call `show()` on the story they want to display after prerendering them.

Enriqe commented:
I've updated the PR so that `prerender()` now appends a story to the player in prerender mode. 

I've also split the show() changes to a separate PR #32615.
